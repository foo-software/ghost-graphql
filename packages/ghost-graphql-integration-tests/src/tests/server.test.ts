import { ApolloServer } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import {
  AuthorsDataSource,
  PagesDataSource,
  PostsDataSource,
  QuerySchema,
  SettingsDataSource,
  TagsDataSource,
} from '@foo-software/ghost-graphql';
import gql from 'graphql-tag';
import mockAuthorResponse from '../mocks/authorResponse';
import mockAuthorsResponse from '../mocks/authorsResponse';
import mockPostResponse from '../mocks/postResponse';
import mockPostsResponse from '../mocks/postsResponse';

// best way of mocking until someone can provide a better example
// https://github.com/apollographql/fullstack-tutorial/issues/90
class AuthorsDataSourceWithMockedGet extends AuthorsDataSource {
  get(): any {
    return {};
  }
}
class PagesDataSourceWithMockedGet extends PagesDataSource {
  get(): any {
    return {};
  }
}
class PostsDataSoureWithMockedGet extends PostsDataSource {
  get(): any {
    return {};
  }
}
class SettingsDataSourceWithMockedGet extends SettingsDataSource {
  get(): any {
    return {};
  }
}
class TagsDataSourceWithMockedGet extends TagsDataSource {
  get(): any {
    return {};
  }
}

const constructTestServer = () => {
  const authorsDataSource = new AuthorsDataSourceWithMockedGet();
  const pagesDataSource = new PagesDataSourceWithMockedGet();
  const postsDataSource = new PostsDataSoureWithMockedGet();
  const settingsDataSource = new SettingsDataSourceWithMockedGet();
  const tagsDataSource = new TagsDataSourceWithMockedGet();

  const server = new ApolloServer({
    schema: QuerySchema,
    dataSources: () => {
      return {
        authorsDataSource,
        pagesDataSource,
        postsDataSource,
        settingsDataSource,
        tagsDataSource,
      };
    },
  });

  return {
    authorsDataSource,
    pagesDataSource,
    postsDataSource,
    settingsDataSource,
    server,
    tagsDataSource,
  };
};

const GET_AUTHOR = gql`
  query author($id: String, $slug: String) {
    author(id: $id, slug: $slug) {
      id
      profileImage
    }
  }
`;

const GET_AUTHORS = gql`
  query authors($limit: Int, $page: Int) {
    authors(limit: $limit, page: $page) {
      edges {
        node {
          id
          profileImage
          slug
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
      }
      meta {
        pagination {
          limit
          next
          page
          pages
          prev
          total
        }
      }
    }
  }
`;

const GET_POST = gql`
  query post($id: String, $slug: String) {
    post(id: $id, slug: $slug) {
      id
      featureImage
      metaDescription
      sendEmailWhenPublished
      slug
    }
  }
`;

const GET_POSTS = gql`
  query posts($limit: Int, $page: Int) {
    posts(limit: $limit, page: $page) {
      edges {
        node {
          id
          featureImage
          metaDescription
          sendEmailWhenPublished
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
      }
      meta {
        pagination {
          limit
          next
          page
          pages
          prev
          total
        }
      }
    }
  }
`;

describe('Queries', () => {
  it('fetches an author item by id', async () => {
    const { authorsDataSource, server } = constructTestServer();

    authorsDataSource.get = jest.fn().mockResolvedValue(mockAuthorResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_AUTHOR,
      variables: { id: 'abc123' },
    });
    expect(res).toMatchSnapshot();
  });

  it('fetches an author item by slug', async () => {
    const { authorsDataSource, server } = constructTestServer();

    authorsDataSource.get = jest.fn().mockResolvedValue(mockAuthorResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_AUTHOR,
      variables: { slug: 'some-slug' },
    });
    expect(res).toMatchSnapshot();
  });

  it('fails to fetch an author when args are missing', async () => {
    const { authorsDataSource, server } = constructTestServer();

    authorsDataSource.get = jest.fn().mockResolvedValue(mockAuthorResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_AUTHOR,
      variables: {},
    });
    expect(res).toMatchSnapshot();
  });

  it('fetches list of authors', async () => {
    const { authorsDataSource, server } = constructTestServer();

    authorsDataSource.get = jest.fn().mockResolvedValue(mockAuthorsResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_AUTHORS,
      variables: { limit: 2, page: 2 },
    });
    expect(res).toMatchSnapshot();
  });

  it('fetches a post item by id', async () => {
    const { postsDataSource, server } = constructTestServer();

    postsDataSource.get = jest.fn().mockResolvedValue(mockPostResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_POST,
      variables: { id: 'abc123' },
    });
    expect(res).toMatchSnapshot();
  });

  it('fetches a post item by slug', async () => {
    const { postsDataSource, server } = constructTestServer();

    postsDataSource.get = jest.fn().mockResolvedValue(mockPostResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_POST,
      variables: { slug: 'welcome' },
    });
    expect(res).toMatchSnapshot();
  });

  it('fails to fetch a post when args are missing', async () => {
    const { postsDataSource, server } = constructTestServer();

    postsDataSource.get = jest.fn().mockResolvedValue(mockPostResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_POST,
      variables: {},
    });
    expect(res).toMatchSnapshot();
  });

  it('fetches list of posts', async () => {
    const { postsDataSource, server } = constructTestServer();

    postsDataSource.get = jest.fn().mockResolvedValue(mockPostsResponse);

    const { query } = createTestClient(server);
    const res = await query({
      query: GET_POSTS,
      variables: { limit: 2, page: 2 },
    });
    expect(res).toMatchSnapshot();
  });
});
