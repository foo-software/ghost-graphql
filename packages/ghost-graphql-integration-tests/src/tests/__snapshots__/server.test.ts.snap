// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fails to fetch a page when args are missing 1`] = `
Object {
  "data": Object {
    "page": null,
  },
  "errors": Array [
    [GraphQLError: either an id or slug needs to be provided],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fails to fetch a post when args are missing 1`] = `
Object {
  "data": Object {
    "post": null,
  },
  "errors": Array [
    [GraphQLError: either an id or slug needs to be provided],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fails to fetch a tag when args are missing 1`] = `
Object {
  "data": Object {
    "tag": null,
  },
  "errors": Array [
    [GraphQLError: either an id or slug needs to be provided],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fails to fetch an author when args are missing 1`] = `
Object {
  "data": Object {
    "author": null,
  },
  "errors": Array [
    [GraphQLError: either an id or slug needs to be provided],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches a page item by id 1`] = `
Object {
  "data": Object {
    "page": Object {
      "createdAt": "2017-07-27T08:31:18.000+00:00",
      "id": "5979a4d6df093500228e9582",
      "slug": "about",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches a page item by slug 1`] = `
Object {
  "data": Object {
    "page": Object {
      "createdAt": "2017-07-27T08:31:18.000+00:00",
      "id": "5979a4d6df093500228e9582",
      "slug": "about",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches a post item by id 1`] = `
Object {
  "data": Object {
    "post": Object {
      "featureImage": "https://demo.ghost.io/content/images/2019/10/welcome-to-ghost.png",
      "id": "5b7ada404f87d200b5b1f9c8",
      "metaDescription": null,
      "sendEmailWhenPublished": false,
      "slug": "welcome",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches a post item by slug 1`] = `
Object {
  "data": Object {
    "post": Object {
      "featureImage": "https://demo.ghost.io/content/images/2019/10/welcome-to-ghost.png",
      "id": "5b7ada404f87d200b5b1f9c8",
      "metaDescription": null,
      "sendEmailWhenPublished": false,
      "slug": "welcome",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches a tag item by id 1`] = `
Object {
  "data": Object {
    "tag": Object {
      "description": null,
      "id": "59799bbd6ebb2f00243a33db",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches a tag item by slug 1`] = `
Object {
  "data": Object {
    "tag": Object {
      "description": null,
      "id": "59799bbd6ebb2f00243a33db",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches an author item by id 1`] = `
Object {
  "data": Object {
    "author": Object {
      "id": "5951f5fca366002ebd5dbef7",
      "profileImage": "//www.gravatar.com/avatar/2bfa103a13c88b5ffd26da6f982f11df?s=250&d=mm&r=x",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches an author item by slug 1`] = `
Object {
  "data": Object {
    "author": Object {
      "id": "5951f5fca366002ebd5dbef7",
      "profileImage": "//www.gravatar.com/avatar/2bfa103a13c88b5ffd26da6f982f11df?s=250&d=mm&r=x",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches list of authors 1`] = `
Object {
  "data": Object {
    "authors": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "5979a779df093500228e958f",
            "profileImage": "https://demo.ghost.io/content/images/2018/10/edgar.jpg",
            "slug": "edgar",
          },
        },
        Object {
          "node": Object {
            "id": "5951f5fca366002ebd5dbef7",
            "profileImage": "//www.gravatar.com/avatar/2bfa103a13c88b5ffd26da6f982f11df?s=250&d=mm&r=x",
            "slug": "ghost",
          },
        },
      ],
      "meta": Object {
        "pagination": Object {
          "limit": 2,
          "next": 3,
          "page": 2,
          "pages": 6,
          "prev": 1,
          "total": 11,
        },
      },
      "pageInfo": Object {
        "hasNextPage": true,
        "hasPreviousPage": true,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches list of pages 1`] = `
Object {
  "data": Object {
    "pages": Object {
      "edges": Array [
        Object {
          "node": Object {
            "createdAt": "2017-07-27T08:31:18.000+00:00",
            "id": "5979a4d6df093500228e9582",
            "slug": "about",
          },
        },
      ],
      "meta": Object {
        "pagination": Object {
          "limit": 2,
          "next": null,
          "page": 1,
          "pages": 1,
          "prev": null,
          "total": 1,
        },
      },
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches list of posts 1`] = `
Object {
  "data": Object {
    "posts": Object {
      "edges": Array [
        Object {
          "node": Object {
            "featureImage": "https://demo.ghost.io/content/images/2019/10/welcome-to-ghost.png",
            "id": "5b7ada404f87d200b5b1f9c8",
            "metaDescription": null,
            "sendEmailWhenPublished": false,
          },
        },
        Object {
          "node": Object {
            "featureImage": "https://demo.ghost.io/content/images/2019/10/writing-posts-with-ghost.png",
            "id": "5b7ada404f87d200b5b1f9c6",
            "metaDescription": null,
            "sendEmailWhenPublished": false,
          },
        },
        Object {
          "node": Object {
            "featureImage": "https://demo.ghost.io/content/images/2019/10/publishing-options.png",
            "id": "5b7ada404f87d200b5b1f9c4",
            "metaDescription": null,
            "sendEmailWhenPublished": false,
          },
        },
        Object {
          "node": Object {
            "featureImage": "https://demo.ghost.io/content/images/2019/10/admin-settings.png",
            "id": "5b7ada404f87d200b5b1f9c2",
            "metaDescription": null,
            "sendEmailWhenPublished": false,
          },
        },
      ],
      "meta": Object {
        "pagination": Object {
          "limit": 15,
          "next": null,
          "page": 1,
          "pages": 1,
          "prev": null,
          "total": 4,
        },
      },
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches list of tags 1`] = `
Object {
  "data": Object {
    "tags": Object {
      "edges": Array [
        Object {
          "node": Object {
            "description": null,
            "id": "59799bbd6ebb2f00243a33db",
          },
        },
        Object {
          "node": Object {
            "description": "Some of the greatest words ever spoken.",
            "id": "5979a779df093500228e958b",
          },
        },
      ],
      "meta": Object {
        "pagination": Object {
          "limit": 2,
          "next": 3,
          "page": 2,
          "pages": 4,
          "prev": 1,
          "total": 8,
        },
      },
      "pageInfo": Object {
        "hasNextPage": true,
        "hasPreviousPage": true,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
